elastalert/ruletypes.py

```diff
@@ -720,6 +721,65 @@ def garbage_collect(self, ts):
                 placeholder.update({self.rules['query_key']: qk})
             self.handle_event(placeholder, 0, qk)
 
+class AdvQueryRule(RuleType):
+    """ A rule that uses a query_string query to perform a search """
+    required_options = frozenset(['alert_field'])
+
+    def __init__(self, *args):
+        super(AdvQueryRule, self).__init__(*args)
+        if 'max_threshold' not in self.rules and 'min_threshold' not in self.rules:
+            raise EAException("AdvQueryRule must have one of either max_threshold or min_threshold")
+        #self.query_string = self.rules.get('query_string')
+        self.rules['aggregation_query_element'] = {"query": ""}
+
+    def run_query(self):
+        # Implement the logic to run the query using the query_string
+        raise NotImplementedError()
+
+    def add_aggregation_data(self, payload):
+        for timestamp, payload_data in payload.items():
+            self.check_matches(payload_data,timestamp)
+    
+    def check_matches(self,data,timestamp):
+        #results=[]
+        for key, value in data.items():
+            if 'buckets' in value:
+                if len(value['buckets']) >=0 :
+                    self.check_matches_recursive(key,value['buckets'],timestamp)
+            else:
+                if self.crossed_thresholds(value['value']):
+                    match={"key":key,"value":value['value']}
+                    self.add_match(match)
+
+    def check_matches_recursive(self,data_key,data_value,timestamp,data_key_value=''):
+        key = data_key
+        if len(data_value) > 0:
```

Comment:
Instead of combining the check_matches & finding key in nested dic logic in same method, Can we have separate method to just do recursive find of a key in nested dict and invoke it by passing the alert_field key and response object?

Reply:
Can do that Rashmi. Generally in check_matches function only thresholds are compared in all rule types. Hence followed the same logic here. Even in Metric aggregation type, elastalert does the same

Reply:
I understand what `check_matches` does. But what I meant is that, we can have separate method to do this nested parsing [https://stackoverflow.com/a/29652561](https://stackoverflow.com/a/29652561) and make use of it.